### PR TITLE
fix: compute critic explained variance globally

### LINKED ISSUE #1247: [Bug]:  Critic explained variance is computed at micro-batch granularity
### Affected component

critic metrics.

### Bug description

The current critic explained variance metric is computed independently for each micro-batch and then averaged across micro-batches. This is mathematically incorrect because explained variance depends on the variance of the full return distribution, which is not additive across micro-batches. As a result, the logged value can be misleading, especially when different micro-batches have different return offsets. The metric should instead be computed at global-batch granularity by accumulating sufficient statistics for returns and prediction errors across micro-batches, then reducing them across actor ranks before logging.

### Configuration YAML file

YAML config file:

You can also paste the full config here.
```
<paste YAML content here>
```

### Log file

You can find the log file in `logs/` folder or the $output_dir/$experiment_name folder (defined in the yaml config) if you are using our example scripts.

Log file:

If you cannot find the log, please provide the full log messages here.
```
<paste log here>
```

### Environment

python -V:
uv pip list:
nvidia-smi:
RLinf version:
Docker image tag: 

### Other reproduction info

_No response_

### Before submitting a new issue...

- [x] Have you checked relevant issues, FAQs (https://rlinf.readthedocs.io/en/latest/rst_source/faq.html), or asked the chatbot at the top right corner of the [documentation page](https://rlinf.readthedocs.io).

### PR DESCRIPTION
**Pull request by ligediaomao**
Fix critic explained variance aggregation by computing it from global sufficient statistics across micro-batches and distributed actor ranks.

## Description

This PR fixes the PPO critic `critic/explained_variance` metric aggregation.

Previously, `critic/explained_variance` could be computed separately for each micro-batch and then averaged. This is mathematically incorrect because explained variance depends on global variance terms:

```text
1 - Var(returns - values) / Var(returns)
```

This PR adds internal sufficient-statistic helpers for critic explained variance, accumulates those hidden statistics across micro-batches and actor ranks, and reconstructs one final global `critic/explained_variance` value before logging.

The public metric name remains unchanged. Hidden internal statistic keys are removed before final metrics are logged.

## Motivation and Context

Fixes #1247.

The previous aggregation could produce misleading critic metrics in PPO training, especially when return distributions differ across micro-batches or distributed actor ranks. This could make the critic look better or worse than it actually is during training.

This change makes the logged `critic/explained_variance` metric mathematically correct for global batches while preserving existing behavior for other scalar metrics.

## How has this been tested?

```bash
PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 .venv/bin/python -m pytest tests/unit_tests/test_critic_metrics.py -q
PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 .venv/bin/python -m pytest tests/unit_tests/test_weight_syncer.py -q
.venv/bin/python -m ruff check rlinf/utils/metric_utils.py rlinf/algorithms/losses.py rlinf/workers/actor/fsdp_actor_worker.py rlinf/workers/actor/async_ppo_fsdp_worker.py rlinf/workers/megatron_worker.py rlinf/workers/critic/megatron_critic_worker.py rlinf/workers/actor/megatron_actor_worker.py rlinf/workers/actor/ma_megatron_actor_worker.py tests/unit_tests/test_critic_metrics.py
git diff --check
```

Results:

```text
tests/unit_tests/test_critic_metrics.py: 5 passed
tests/unit_tests/test_weight_syncer.py: 29 passed, 1 skipped
ruff: passed
git diff --check: passed
```

## Additional information (optional, e.g., figures and logs):

The new regression test covers a case where averaging per-micro-batch explained variance differs from the correct global-batch value, and verifies that summed sufficient statistics reproduce the full-batch explained variance.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by zoeyuchao** [comment]
@ligediaomao thanks for pointing out this issue. Could you further solve the error in doc ci test so we can merge this pr asap?

**Comment by ligediaomao** [comment]
> @ligediaomao thanks for pointing out this issue. Could you further solve the error in doc ci test so we can merge this pr asap?

Thanks! I’ll look into the doc CI failure and fix it as soon as possible.

**Comment by ligediaomao** [comment]
> @ligediaomao thanks for pointing out this issue. Could you further solve the error in doc ci test so we can merge this pr asap?

Thanks! I have fixed the doc CI issue.

File: rlinf/algorithms/losses.py [hunk 1]
```diff
@@ -360,30 +365,24 @@ def compute_ppo_critic_loss(
...
+    explained_variance = compute_critic_explained_variance_from_stats(
+        explained_variance_stats
+    )
 
     # Compile metrics for logging
     metrics_data = {
         "critic/value_loss": value_loss.detach(),
         "critic/value_clip_ratio": value_clip_ratio.detach(),
-        "critic/explained_variance": explained_variance.detach(),
+        CRITIC_EXPLAINED_VARIANCE_KEY: explained_variance.detach(),
```
Comment (qurakchin) [inline rlinf/algorithms/losses.py]: It seems that all the CRITIC_EXPLAINED_VARIANCE_KEY will be re-compute in actor workers. So is this redundant?

File: rlinf/utils/metric_utils.py [hunk 2]
```diff
@@ -20,11 +20,114 @@
...
+    CRITIC_EXPLAINED_VARIANCE_RETURNS_SQ_SUM_KEY,
+    CRITIC_EXPLAINED_VARIANCE_ERRORS_SUM_KEY,
+    CRITIC_EXPLAINED_VARIANCE_ERRORS_SQ_SUM_KEY,
+)
 
 def compute_split_num(num, split_num):
     return math.lcm(num, split_num) // split_num
 
 
+def compute_critic_explained_variance_stats(
```
Comment (qurakchin) [inline rlinf/utils/metric_utils.py:55]: in rlinf/workers/agent/agent_loop.py, i used a post_process_metric with a prefix in one of ("__sum__/", "__max__/", "__min__/", "__mean__/") to generally do automatic metrics reduce. can the metric_utils use this style?

**Comment by zoeyuchao** [comment]
@ligediaomao Hi, it seems the pr title doesn't pass the ci test and the code doesn't pass lint check. please fix these errors.

File: rlinf/algorithms/losses.py (hunk 1 above)
Comment (ligediaomao) [reply rlinf/algorithms/losses.py]: > It seems that all the CRITIC_EXPLAINED_VARIANCE_KEY will be re-compute in actor workers. So is this redundant?

Thanks for pointing this out. I removed the intermediate public critic/explained_variance; it is now only reconstructed after reducing the sufficient stats.

File: rlinf/utils/metric_utils.py (hunk 2 above)
Comment (ligediaomao) [reply rlinf/utils/metric_utils.py:55]: > in rlinf/workers/agent/agent_loop.py, i used a post_process_metric with a prefix in one of ("**sum**/", "**max**/", "**min**/", "**mean**/") to generally do automatic metrics reduce. can the metric_utils use this style?

Thanks, updated. The hidden stats now use the existing __sum__/... reduction-prefix style.

**Comment by ligediaomao** [comment]
> @ligediaomao Hi, it seems the pr title doesn't pass the ci test and the code doesn't pass lint check. please fix these errors.

Thanks for the reminder. I fixed the PR title and ruff formatting issues, and pushed the updates.
